File: 1/AI1110/Random-Numbers/codes/5.7.py
# Plots theoretical and empirical P_e with respect to A from 0 to 10 dB

# Name: Ankit Saha
# Roll number: AI21BTECH11004

import numpy as np
import matplotlib.pyplot as plt
import mpmath as mp
import scipy

data = np.loadtxt('5.7.dat', dtype='double')
NUM = len(data)
A = np.linspace(0, 10, 41)
p_e_0 = data[:, 0]
p_e_1 = data[:, 1]
p_e = 0.5 * p_e_0 + 0.5 * p_e_1

def Q_function(x):
    return 0.5 * mp.erfc(x/(2**0.5))

vec_p_e = scipy.vectorize(Q_function, otypes=[float])

plt.plot(A, p_e, 'o', label='Empirical')
plt.plot(A, vec_p_e(A), label='Theoretical')
plt.grid()
plt.legend()
plt.xlabel('$A$')
plt.ylabel('$P_e$')
plt.savefig('../figs/5.7.png')
plt.show()

File: 1/AI1110/Random-Numbers/codes/5.7.c
// Writes to a file the probabilities P_{e|0} and P_{e|1} for different values of A

// Name: Ankit Saha
// Roll number: AI21BTECH11004

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "header.h"

#define NUM 1000000

int main() {
	FILE *fp = fopen("5.7.dat", "w");
	for (double A = 0; A <= 10; A += 0.25) {
		int e_0 = 0;
		int e_1 = 0;
		generateY("y.dat", NUM, A);
		double *y_data = readData("y.dat", NUM);
		double *x_data = readData("x.dat", NUM);
		for (int i = 0; i < NUM; i++) {
			if (y_data[i] < 0 && x_data[i] > 0) {
				e_0++;
			} else if (y_data[i] > 0 && x_data[i] < 0) {
				e_1++;
			}
		}	
		double p_e_0 = (double)e_0 / NUM;
		double p_e_1 = (double)e_1 / NUM;
		fprintf(fp, "%lf %lf\n", p_e_0, p_e_1);
	}
	fclose(fp);
	return 0;	
}
